pages/api/system/reload.js

```diff
@@ -0,0 +1,279 @@
+import fs from "fs";
+import path from "path";
+
+import connectMongo from "@config/mongo";
+import logger from "@config/logger";
+import Profile from "@models/Profile";
+import Link from "@models/Link";
+
+export default async function handler(req, res) {
+  if (
+    req.method !== "GET" ||
+    req.query.secret !== process.env.LINKFREE_RELOAD_SECRET
+  ) {
+    logger.error(
+      `attempt to load profile json but security check failed: "${req.query.secret}"`
+    );
+    return res.status(401).json({ error: "ONLY system calls allowed" });
+  }
+  await connectMongo();
+
+  // 1. load all profiles
+  const basicProfiles = findAllBasic();
+  const fullProfiles = basicProfiles.map((profile) =>
+    findOneByUsernameFull(profile)
+  );
+
+  // 2. save basic profiles to database
+  // only if `source` is not `database` (this will be set when using forms)
+  fullProfiles.map(async (profile) => {
```

Comment:
This section need to be inside a `Promise.all()` which you will need to await. Otherwise the function will return before the code is finished running.

Reply:
I thought the promise all is only required if I didn't `await` for each one and wanted them done in parallel? My understanding is at the moment it does one at a time and waits for each one t complete before moving to the next one

Reply:
For example (simplified)...

**Current sequentially**
```js
fullProfiles.map(async (profile) => 
   await Profile.findOne({
      username: profile.username,
   });
);
```

VS

**In parallel**

```js
await Promise.all(fullProfiles.forEach((profile) => 
   Profile.findOne({
      username: profile.username,
   });
));
```

Reply:
First off, all async functions return a promise.

Using `array.map()` or `array.forEach()` with an async callback function means they can run concurrently (they won't truly run in parallel as node is single threaded by default).

From my testing, if I limit the fullProfiles array to 100 users, everything works fine.
But when I use the full array, I get the same errors the tests get. 
I believe this is caused by database requests backing up and all the promises are returned before the functions are complete.

Thus, using `Promise.all()` will ensure everything is complete before returning.

Reply:
> all async functions return a promise.

oh that is a good point 👍 


Thanks for testing, I will update 👍